[ARMx32] Problem when compiling on Raspberry Pi 4

worrysome warning when compiling on raspberry pi 4

~~~
In file included from /home/pi/Kratos/kratos/includes/key_hash.h:19,
                 from /home/pi/Kratos/kratos/includes/global_pointer.h:20,
                 from /home/pi/Kratos/kratos/containers/global_pointers_vector.h:26,
                 from /home/pi/Kratos/kratos/includes/deprecated_variables.h:43,
                 from /home/pi/Kratos/kratos/sources/deprecated_variables.cpp:19:
/home/pi/Kratos/kratos/includes/dof.h: In instantiation of ‘class Kratos::Dof<double>’:
/home/pi/Kratos/kratos/includes/key_hash.h:318:46:   required from here
/home/pi/Kratos/kratos/includes/dof.h:480:34: warning: width of ‘Kratos::Dof<double>::mEquationId’ exceeds its type
     EquationIdType mEquationId : 48;
                                  ^~
In file included from /home/pi/Kratos/kratos/includes/key_hash.h:19,
                 from /home/pi/Kratos/kratos/includes/global_pointer.h:20,
                 from /home/pi/Kratos/kratos/containers/global_pointers_vector.h:26,
                 from /home/pi/Kratos/kratos/includes/c2c_variables.h:34,
                 from /home/pi/Kratos/kratos/sources/c2c_variables.cpp:22:
/home/pi/Kratos/kratos/includes/dof.h: In instantiation of ‘class Kratos::Dof<double>’:
/home/pi/Kratos/kratos/includes/key_hash.h:318:46:   required from here
/home/pi/Kratos/kratos/includes/dof.h:480:34: warning: width of ‘Kratos::Dof<double>::mEquationId’ exceeds its type
     EquationIdType mEquationId : 48;
~~~

@pooyan-dadvand this looks very worrysome

--- a/kratos/includes/define.h
+++ b/kratos/includes/define.h
@@ -25,7 +25,7 @@
 #include "includes/shared_pointers.h"
 #include "includes/exception.h"
 
-
+// Defining the OS
 #if defined(__linux__) || defined(__linux) || defined(linux) || defined(__gnu_linux__)
     #define KRATOS_COMPILED_IN_LINUX
 
@@ -36,6 +36,21 @@
     #define KRATOS_COMPILED_IN_WINDOWS
 #endif
 
+// Defining the architecture (see https://sourceforge.net/p/predef/wiki/Architectures/)
+// Check Windows
+#if defined(_WIN32) || defined(_WIN64)
+   #if defined(_WIN64)
+     #define ENV64BIT
+   #else
+     #define ENV32BIT
+  #endif
+#else // It is POSIX (Linux, MacOSX, BSD...)
+  #if defined(__x86_64__) || defined(__ppc64__) || defined(__aarch64__)
+    #define ENV64BIT
+  #else // This includes __arm__ and __x86__
+    #define ENV32BIT
+  #endif
+#endif
 
 //-----------------------------------------------------------------
 //

--- a/kratos/includes/dof.h
+++ b/kratos/includes/dof.h
@@ -477,7 +477,7 @@ private:
     int mIndex : 6;
 
     /** Equation identificator of the degree of freedom */
-#if defined (__arm__)
+#if defined (__arm__) // Required to avoid overflow on ARM systems (32 bit)
     EquationIdType mEquationId : 32;
 #else
     EquationIdType mEquationId : 48;

--- a/kratos/includes/dof.h
+++ b/kratos/includes/dof.h
@@ -477,7 +477,7 @@ private:
     int mIndex : 6;
 
     /** Equation identificator of the degree of freedom */
-#if defined (__arm__) // Required to avoid overflow on ARM systems (32 bit)
+#if ENV32BIT // Required to avoid overflow on 32 bit systems
     EquationIdType mEquationId : 32;
 #else
     EquationIdType mEquationId : 48;

--- a/kratos/includes/dof.h
+++ b/kratos/includes/dof.h
@@ -478,7 +478,7 @@ private:
 
     /** Equation identificator of the degree of freedom */
 #if defined (__arm__)
-    EquationIdType mEquationId : 16;
+    EquationIdType mEquationId : 32;
 #else
     EquationIdType mEquationId : 48;
 #endif

--- a/kratos/includes/dof.h
+++ b/kratos/includes/dof.h
@@ -477,7 +477,7 @@ private:
     int mIndex : 6;
 
     /** Equation identificator of the degree of freedom */
-#if ENV32BIT // Required to avoid overflow on 32 bit systems
+#ifdef ENV32BIT // Required to avoid overflow on 32 bit systems
     EquationIdType mEquationId : 32;
 #else
     EquationIdType mEquationId : 48;

--- a/kratos/includes/dof.h
+++ b/kratos/includes/dof.h
@@ -477,7 +477,11 @@ private:
     int mIndex : 6;
 
     /** Equation identificator of the degree of freedom */
+#if defined (__arm__)
+    EquationIdType mEquationId : 16;
+#else
     EquationIdType mEquationId : 48;
+#endif
 
     /** A pointer to nodal data stored in node which is corresponded to this dof */
     NodalData* mpNodalData;

